src/censored.jl

```diff
@@ -0,0 +1,168 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+
+function censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+    l <= u || error("the lower bound must be less or equal than the upper bound")
+    Censored(d, l, u)
+end
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+
+params(d::Censored) = tuple(params(d.uncensored)..., d.lower, d.upper)
+partype(d::Censored) = partype(d.uncensored)
+Base.eltype(::Type{<:Censored{D} } ) where {D} = eltype(D)
+
+### range and support
+
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || isfinite(d.lower)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    return d.lower <= x <= d.upper && insupport(d.uncensored, x)
+end
+
+### evaluation
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if x == lower
+        cdf(d0, x)
+    elseif x == upper
+        uccdf = ccdf(d0, x)
+        if value_support(typeof(d)) === Discrete
```

Comment:
This could possibly go wrong if you censor a continuous distribution with atoms ("continuous is the weak sense" we are converging to).

Reply:
Fair point, what I need to do is subtract out the atomic probability mass at that point (so the ccdf includes the bound). How would you recommend doing that?

Reply:
Exactly, introducing continuous distributions with atoms we are creating a situation where for `<:Continuous` distributions `pdf` may return a density or an atomic mass and we have no yet established a way of querying if a value is an atom or getting explicity the probability of something. Here I almost think we could make use of a dedicated function `probability(d, x)` giving the probability of `P(X = x)` with fallback `= 0` for continuous and `pdf(d, x)` for discrete

Reply:
Something like that would be useful. For us to assume it is always correct, we would need to make it part of the API. `Truncated` would benefit also from this: https://github.com/JuliaStats/Distributions.jl/blob/7206033e2113ded8758bcaa408deadb5156500ec/src/truncate.jl#L24-L29

@devmotion, thoughts?

Reply:
Yep, I already thought that this could be useful for `truncated` as well. I'm not completely happy with`probability`, it seems a bit too general and unspecific for my taste. Maybe `probabilitymass` would be better, or maybe we should just resurrect `pmf` and `logpmf` but this time for both `DiscreteDistribution` and `ContinuousDistribution` (the default would be 0 for `ContinuousDistribution`).

Reply:
I am happy with `(log)pmf`

Reply:
Probably should be done outside of this PR. Would someone like to create an issue?

Reply:
Yes we can proceed here because `censored` of `Censored` is `Censored` and therefore there isn't yet a case where this goes wrong.